packages/@aws-cdk/aws-iam/docs/policy-merging.als

```diff
@@ -0,0 +1,116 @@
+pred applies[s: Statement, req: Request] {
+  req.principal in s.principal or req.principal not in s.notPrincipal
+  req.action in s.action or req.action not in s.notAction
+  req.resource in s.resource or req.resource not in s.notResource
+}
+
+// Whether or not to allow the given request according to the given statements
+//
+// A request is allowed if there's at least one statement allowing it and
+// no statements denying it.
+pred allow[req: Request, ss: some Statement] {
+  some s: ss | applies[s, req] and s.effect = Allow
+  no s: ss | applies[s, req] and s.effect = Deny
+}
+
+---------------------------------------------------------
+-- Statement merging
+
+// Helpers to assert that certain fields are the same
+let sameAction[a, b] = { a.action = b.action and a.notAction = b.notAction }
+let sameResource[a, b] = { a.resource = b.resource and a.notResource = b.notResource }
+let samePrincipal[a, b] = { a.principal = b.principal and a.notPrincipal = b.notPrincipal }
+
+// Assert that m is the merged version of a and b
+//
+// This encodes the important logic: the rules of merging.
+pred merged[a: Statement, b: Statement, m: Statement] {
+  m.effect = a.effect and m.effect = b.effect 
+  {
+    // Writing 'some a.action', 'some b.action' here is critical
+    // This asserts we are dealing with a POSITIVE action and not
+    // a notAction (that's excluded by the fact on Statement that only
+    // one of the two can be set.
+    //
+    // We can show that we have a bug if you uncomment the next line.
+    some a.action and some b.action // Excludes notAction
+
+    // The constraint on notAction here is not necessary, because
+    // in practice it must always the empty set, but it is necessary to
+    // correctly show the bug in action if the previous line is commented out.
+    m.action = a.action + b.action and m.notAction = a.notAction + b.notAction
+    sameResource[a, b] and sameResource[b, m]
+    samePrincipal[a, b] and samePrincipal[b, m]
+  } or {
+    some a.resource and some b.resource // Excludes notResource
+    m.resource = a.resource + b.resource and m.notResource = a.notResource + b.notResource
+    sameAction[a, b] and sameAction[b, m]
+    samePrincipal[a, b] and samePrincipal[b, m]
+  } or {
+    some a.principal and some b.principal // Excludes notPrincipal
+    m.principal = a.principal + b.principal and m.notPrincipal = a.notPrincipal + b.notPrincipal
+    sameAction[a, b] and sameAction[b, m]
+    sameResource[a, b] and sameResource[b, m]
+  }
+}
+
+// Maximum merging of all statements in orig
+pred maxMerged[orig: some Statement, merged: some Statement] {
+}
+
+// There are no statements so that if they are merged, 
+// the evaluation of the separate statements is different than the evaluation
+// of the merged statement
+check merging_does_not_change_evaluation {
+  no a: Statement, b: Statement, m: Statement, r: Request {
+    merged[a, b, m]
+    allow[r, a + b] iff not allow[r, m]
+  }
+} for 5
```

Comment:
I’ll be happy to explain this in detail btw

Reply:
Maybe good topic for a team time even?

Reply:
It would be nice if we could ensure they are side-by-side able... Like if we could annotate the TypeScript with the Alloy model pieces, and then have a tool that extracts the alloy model from the code.

This way would reduce the likelihood that one side is changed without the other side being changed to match; and would hopefully make it easier to validate that the code is expressing the same thing as the model.

Now - this is obviously work for a different PR, that is already more than enough things to look at in this one...